src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,30 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (CacheDictionary.ContainsKey(key))
+                return CacheDictionary[key];
+
             foreach (var entity in entities) {
+                if (CacheValues.Contains(entity))
+                    continue;
                 var keyIdx = 0;
                 foreach (var value in extractor.ExecutePiecewise(entity)) {
                     if (!value.Equals(key[keyIdx++])) {
                         goto nextEntity;
                     }
                 }
+                CacheDictionary[key] = entity;
+                CacheValues.Add(entity);
                 return entity;
                 nextEntity: continue;
             }
 
             // This code should be unreachable
             throw new ApplicationException($"No entity found to match key: {string.Join(", ", key)}");
         }
+
+        private static readonly Dictionary<DBData, object> CacheDictionary = new();
+        private static readonly HashSet<object> CacheValues = new();
```

Comment:
What are your thoughts on the equality comparator that gets used for the `HashSet`? Conceptually, we want to use an identify (i.e. `ReferenceEquals(lhs, rhs)`) check, but the default (I _think_) is actually to use `object.Equals(rhs)` instead. My biggest concern is this delegating to a user-defined `Equals` function that is either (A) expensive or (B) treats two Entities with different primary keys as being equal. Obviously, the latter is likely indicative of some issue on the user's end . . . I'm not sure if we should be looking to mitigate it, though.

Reply:
I agree that we shouldn't trust the user's `Equals` function, but I'm not sure that we have any actual references to work with - the `entities` parameter to `ByKey` might contain equivalent data stored in different objects during different calls of this method (perhaps they were computed using different but equivalent functions, or the work was duplicated by the user, etc.). You're right, `HashSet.Contains` uses the object's `Equals` function, which likely is insufficient, but I'm unsure about the best alternative if we don't have meaningful references.

Reply:
There's also another potential issue: If we're using `ReferenceEquals`, then our cache might become invalid if a `key` is calculated in multiple locations, which would mean that the keys have different hash codes. Perhaps we could use a more explicit `DBData` type which calculates its hash code based on values rather than reference?

Reply:
Alright, finally getting back to this.

I was specifically referring to using `ReferenceEquals` _only_ for the `HashSet` of Entity objects, not for the key. I 100% agree that the key is going to have to be an equality check, since each time we reflect over the source object to obtain the key we'll end up with a brand new `DBData` reference; changing that would require fundamental changes to the Extraction Layer, which I don't think is appropriate.

As far as
> the `entities` parameter to `ByKey` might contain equivalent data stored in different objects during different calls of this method
. . . I think this is technically true, but unlikely in practice. But even if it were true, using `ReferenceEquals` would just result in false negatives (e.g. `A` and `B` are supposed to be the same Entity but `!ReferenceEquals(A, B)`). Then the new instance would be added to the `HashSet` for the next round. I still think that's fine, and I think it makes sense to do so that we avoid relying on the user's `Equals` (which will fall back to `ReferenceEquals` if not implemented explicitly _anyway_).